### ISSUE TITLE
Create brownie tests for DaoCross 0.1

### ISSUE THREAD

**Issue by tamriq** [sorrentum/sorrentum#136]
Once https://github.com/sorrentum/sorrentum/pull/125 is merged, we need to cover the whole DaoCross functionality with unit tests.
I already created one simple test file to check if matchOrders works as expected. Also I created mock contracts for erc20 and for price aggregator. Feel free to extend my testing code to cover specific cases.
Right now DaoCross works with erc20<=>eth orders.
We need to pay special attention to testing the Min Heap implementation (OrderMinHeap.sol), let's find some corner cases and test them.

**Comment by samarth9008** [comment]
@tamriq 

```
function insert(Heap storage self, Order storage order) internal {
        if (self.size == self.data.length) {
            self.data.push();
        }
        uint32 _index = self.size++;
        while (_index > 0 && self.data[parent(_index)].quantity < order.quantity) {
            self.data[_index] = self.data[parent(_index)];
            _index = parent(_index);
        }
        self.data[_index] = order;
    }
```

As far as my understanding, here we are implementing min-heap based on the quantity of the specific orders. So the order with minimun quantity will be the root element. While inserting new order to the heap, we first check the size of heap and then increases the value of index accordingly. While loop will run until `_index` value is greater than 0 and the quantity of parent order is less than the quantity of the order being inserted. So if the parent's quantity is less than the current order's quantity than no swap is needed to satisfy the min-heap property. So shouldn't the correct comparison be like `self.data[parent(_index)].quantity > order.quantity`? 

I am not sure if I am right. Please correct me if I am wrong.

**Comment by tamriq** [comment]
@samarth9008 You are right! Glad u noticed this, Samarth, good job! Feel free to create a PR with a fast fix

**Comment by samarth9008** [comment]
@tamriq 

```
if (leftChild < self.data.length && self.data[leftChild].quantity > self.data[min].quantity) {
                min = leftChild;
            }
            if (rightChild < self.data.length && self.data[rightChild].quantity > self.data[min].quantity) {
                min = rightChild;
            }
```

I found similar issue in `heapify` function. There should be `<` comparison in both the `if` conditions. Right?

**Comment by tamriq** [comment]
Exactly!

**Comment by samarth9008** [comment]
@tamriq 

In the `sellOrder` function, whenever user will place sell order, we will collect tokens from them using   `baseToken.transferFrom(msg.sender, address(this), _quantity);` so the tokens will be stored in this contract. Am I right about this?

If so, then in the `matchOrders` if the `order.limitPrice > clearingPrice` for sell order, the order will not be placed and it will be transfer to the `returnTransfer` variable of `transfer` struct. 

```
returnTransfer = Transfer(
                        order.baseToken,
                        order.quantity,
                        order.walletAddress,
                        order.depositAddress,
                        true
                    );
```
From `returnTransfer`, `order.walletAddress` which represents to `msg.sender` according to the `Order` struct, will be the address from which tokens will be transfered to ` order.depositAddress` as far as my understanding about this. However, the user of sell order have already submitted their tokens to us, how will they recieve back their token from `order.walletAddress`. Please correct me if I am wrong about the whole process.

**Comment by samarth9008** [comment]
@tamriq Is there any update on this?

**Comment by tamriq** [comment]
FYI Samarth, I created a separate issue for testing OrderMinHeap in isolation, so u can concentrate on DaoCross functionality only. https://github.com/sorrentum/sorrentum/issues/165

**Comment by samarth9008** [comment]
@tamriq I understand we are using just one array for all transfers and return transfers but I am still not sure. Lets suppose the user wants to sell 3ADA and suppose 1 ADA = 1 ETH, so the order notation will be 3 ADA / 1 ETH and the user will recieve 3 ETH. The DaoCross contract will recieve the 3 ADA from user and the `sellOrder` function will create an `order` like

```
orders.length = 1,
msg.sender = address of the user calling the `sellOrder`,
_baseToken = address of ADA,
address(0x0), = address of ETH
_quantity, = 3
_limitPrice, = 1
false,
_depositAddress, = address where ETH will be deposited from any `buy` order
block.timestamp
```

Now in the `matchOrder` function, lets suppose the `clearingPrice` is less than the `limitPrice`, so the order will not be placed, we will return 3 ADA to the user and move the order to `returnTransfer`. The `returnTransfer` will look like
```
returnTransfer = Transfer(
                        order.baseToken = address of ADA,
                        order.quantity = 3,
                        from order.walletAddress = msg.sender(), // I have doubt in this line specifically
                        to order.depositAddress = deposit Address mentioned by user,
                        true
                    );
```
So how will we send back tokens to the user from `msg.sender` to `depositAddress` since the user have already transfer their tokens to us using 
`baseToken.transferFrom(msg.sender, address(this), _quantity);`? 

Or does this line `baseToken.transfer(transfer.to, transfer.amount);` in `onSwapTime` function will transfer 3 ADA back to user from this DaoCross contract?

**Comment by tamriq** [comment]
> Or does this line baseToken.transfer(transfer.to, transfer.amount); in onSwapTime function will transfer 3 ADA back to user from this DaoCross contract?

Exactly! I understand this is tricky so let's define the difference between `transferFrom` and `transfer`. Both functions serve a similar purpose — to transfer tokens from one account to another — but they have a key difference:
- The `transfer` function is typically used to send tokens directly from the calling account (from which the function was called, DaoCross contract in our case) to a specified recipient. It requires that the calling account has enough tokens to cover the transfer amount. In our case when we use transfer we already have the tokens from users. The function signature for transfer looks like this:
`function transfer(address recipient, uint256 amount) public returns (bool);`

- The `transferFrom` function, on the other hand, is used to transfer tokens from one account to another on behalf of the token owner. Before using transferFrom, the token owner must first approve the spender to transfer a specific amount of tokens on their behalf using the approve function. That's why in tests we execute this line `base_token.approve(daocross, 2 * 10**18, {"from": accounts[4]})` before creating the sell order. In fact this action should be done by the user before they submit sell order. The function signature for `transferFrom` looks like this:
`function transferFrom(address sender, address recipient, uint256 amount) public returns (bool);`
When using `transferFrom`, the tokens are moved from the specified sender address to the recipient address, as long as the spender (msg.sender) has been granted approval by the sender for the specified amount.

**Comment by samarth9008** [comment]
Okay! Got it! Thank you so much for detailed clarification. The whole flow is clear now.
